### PR TITLE
[GOAL2-558] main net s3 bucket strategy

### PR DESCRIPTION
**Pull request by egieseke**
## Summary

Removed redundant  s3Helper.go files from updater and logger.
Replaced remaining references to S3_UPLOAD_ID and S3_UPLOAD_SECRET with AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY.
 
## Test Plan

Tested building with Travis and with Algonet.

### CODE REVIEW

File: logging/s3Helper.go [hunk 1]
```diff
@@ -31,22 +32,25 @@ type s3Helper struct {
 	bucket  string
 }
 
+const s3UploadBucket = "algorand-uploads"
```
Comment (Karmastic) [inline logging/s3Helper.go]: UploadBucket and Region can have defaults but need to be overridable from the environment.

File: scripts/create_and_deploy_recipe.sh [hunk 2]
```diff
@@ -18,8 +18,8 @@
 
 set -e
 
-if [[ "${S3_UPLOAD_ID}" = "" || "${S3_UPLOAD_SECRET}" = "" || "${S3_UPLOAD_BUCKET}" = "" ]]; then
-    echo "You need to export S3_UPLOAD_ID, S3_UPLOAD_SECRECT and S3_UPLOAD_BUCKET for this to work"
+if [[ "${AWS_ACCESS_KEY_ID}" = "" || "${AWS_SECRET_ACCESS_KEY}" = "" ]]; then
```
Comment (Karmastic) [inline scripts/create_and_deploy_recipe.sh]: Restore requirement for specifying S3_UPLOAD_BUCKET.

File: scripts/upload_config.sh [hunk 3]
```diff
@@ -2,7 +2,6 @@
 set -e
 
 # upload_config.sh - Archives and uploads a netgoal configuration package from a specified directory
-#           NOTE: Will only work if you have the required S3_UPLOAD_ID/SECRET vars set
```
Comment (Karmastic) [inline scripts/upload_config.sh]: Don't these notes still apply?

File: test/testdata/deployednettemplates/recipes/geo-test-small/Makefile [hunk 4]
```diff
@@ -0,0 +1,12 @@
+PARAMS=-w 80 -R 1 -N 4 -n 8 --node-template node.json --relay-template relay.json
```
Comment (Karmastic) [inline test/testdata/deployednettemplates/recipes/geo-test-small/Makefile]: These recipe files do not belong in this PR, do they?

File: util/s3/s3Helper.go [hunk 5]
```diff
@@ -34,6 +33,9 @@ import (
 	"github.com/aws/aws-sdk-go/service/s3/s3manager"
 )
 
+const algorandReleasesBucketName = "algorand-releases"
```
Comment (Karmastic) [inline util/s3/s3Helper.go]: Same comments as above.  Also, we should take this opportunity to modify updater to use the shared version of the s3helper code - we don't need all the duplicate code.  I didn't have time/energy when I added this to modify updater to use it.

File: logging/s3Helper.go (hunk 1 above)
Comment (egieseke) [reply logging/s3Helper.go]: OK, updated to allow overriding with defaults with S3_UPLOAD_BUCKET and S3_REGION environment variables.

File: scripts/create_and_deploy_recipe.sh (hunk 2 above)
Comment (egieseke) [reply scripts/create_and_deploy_recipe.sh]: Restored the requirement for S3_UPLOAD_BUCKET.

File: util/s3/s3Helper.go (hunk 5 above)
Comment (egieseke) [reply util/s3/s3Helper.go]: OK, consolidated the 2 files into 1 within util/s3.

File: test/testdata/deployednettemplates/recipes/geo-test-small/Makefile (hunk 4 above)
Comment (egieseke) [reply test/testdata/deployednettemplates/recipes/geo-test-small/Makefile]: I am using this to test algonet after these changes, but I will stash it before merging.

File: util/s3/s3Helper.go [hunk 6]
```diff
@@ -61,19 +63,56 @@ func loadS3Keys(keyFile string) (keys s3Keys, err error) {
...
+	if !found {
+		bucketName = s3DefaultUploadBucket
+	}
+	return
+}
+
+func getS3ReleaseBucket() (bucketName string) {
+	bucketName, found := os.LookupEnv("S3_RELEASE_BUCKET")
+	if !found {
+		bucketName = algorandReleasesBucketName
```
Comment (Karmastic) [inline util/s3/s3Helper.go]: for consistent naming, change to s3DefaultReleaseBucket

File: util/s3/s3Helper.go [hunk 7]
```diff
@@ -61,19 +63,56 @@ func loadS3Keys(keyFile string) (keys s3Keys, err error) {
...
 }
 
 // MakeS3SessionForUpload returns an s3.Helper for the default algorand S3 bucket - for uploading
 func MakeS3SessionForUpload() (helper Helper, err error) {
 	awsID, _ := os.LookupEnv("AWS_ACCESS_KEY_ID")
 	awsKey, _ := os.LookupEnv("AWS_SECRET_ACCESS_KEY")
-	awsBucket := algorandBuildsBucketName
-	if awsID == "" || awsKey == "" || awsBucket == "" {
+	awsBucket := getS3UploadBucket()
+	if awsID == "" || awsKey == "" {
```
Comment (Karmastic) [inline util/s3/s3Helper.go]: still need to check for awsBucket == "" -- you can have a blank environment variable.

File: test/testdata/deployednettemplates/recipes/geo-test-small/Makefile (hunk 4 above)
Comment (Karmastic) [reply test/testdata/deployednettemplates/recipes/geo-test-small/Makefile]: ok - thanks.  please don't forget :)

File: util/s3/s3Helper.go (hunk 6 above)
Comment (egieseke) [reply util/s3/s3Helper.go]: OK, good thinking.

File: util/s3/s3Helper.go (hunk 7 above)
Comment (egieseke) [reply util/s3/s3Helper.go]: OK, added a check for empty bucket name.

File: test/testdata/deployednettemplates/recipes/geo-test-small/Makefile (hunk 4 above)
Comment (egieseke) [reply test/testdata/deployednettemplates/recipes/geo-test-small/Makefile]: Removed

File: cmd/nodecfg/download.go [hunk 8]
```diff
@@ -38,7 +38,7 @@ func downloadAndExtractConfigPackage(channel string, targetDir string) (err erro
 }
 
 func downloadConfigPackage(channelName string, targetDir string) (packageFile string, err error) {
-	s3, err := s3.MakeS3SessionForDownload()
+	s3, err := s3.MakePublicS3SessionForDownload()
```
Comment (Karmastic) [inline cmd/nodecfg/download.go]: Nodecfg should be downloading from a private bucket -- these shouldn't be for public consumption.

File: util/s3/s3Helper.go [hunk 9]
```diff
@@ -59,37 +64,121 @@ func loadS3Keys(keyFile string) (keys s3Keys, err error) {
...
+	creds := credentials.NewStaticCredentials(awsID, awsKey, "")
+	return makeS3Session(creds, awsBucket)
+}
+
+func makeTravisS3SessionForUpload() (helper Helper, err error) {
+	return makeLocalS3SessionForUpload(getS3ReleaseBucket())
+}
+
+// MakePublicS3SessionForDownload returns an s3.Helper for the default algorand S3 bucket - for downloading
+func MakePublicS3SessionForDownload() (helper Helper, err error) {
```
Comment (Karmastic) [inline util/s3/s3Helper.go]: s3Helper should have no specializations - it should be generic, taking credentials, region, and bucket names.  We can export constants for the default credentials and/or buckets but that logic should be external.

Comment (winder) [reply util/s3/s3Helper.go]: I would prefer doing that in a followup PR, consolidating the 3x `s3Helper.go` files already has a lot of moving parts without refactoring them at the same time.

**Comment by winder** [comment]
We'll need to update Travis, it's using the `S3_UPLOAD_*` parameters for the nightly build deployment.

File: scripts/create_and_deploy_private_version.sh [hunk 10]
```diff
@@ -21,8 +21,8 @@
 
 set -e
 
-if [[ "${S3_UPLOAD_ID}" = "" || "${S3_UPLOAD_SECRET}" = "" || "${S3_UPLOAD_BUCKET}" = "" ]]; then
-    echo "You need to export S3_UPLOAD_ID, S3_UPLOAD_SECRECT and S3_UPLOAD_BUCKET for this to work"
+if [[ "${AWS_ACCESS_KEY_ID}" = "" || "${AWS_SECRET_ACCESS_KEY}" = "" || "${S3_UPLOAD_BUCKET}" = "" ]]; then
+    echo "You need to export AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY and S3_UPLOAD_BUCKET for this to work"
```
Comment (winder) [inline scripts/create_and_deploy_private_version.sh]: I think that when `-b` is provided we don't need S3_UPLOAD_BUCKET

File: scripts/create_and_deploy_recipe.sh [hunk 11]
```diff
@@ -93,5 +93,5 @@ ${SRCPATH}/scripts/upload_config.sh "${ROOTDIR}" "${CHANNEL}"
 
 if [ "${NO_DEPLOY}" = "" ]; then
     # Now generate a private build using our custom genesis.json and deploy it to S3 also
-    ${SRCPATH}/scripts/deploy_private_version.sh -c "${CHANNEL}" -f "${ROOTDIR}/genesisdata/genesis.json" -n "${NETWORK}"
+    ${SRCPATH}/scripts/deploy_private_version.sh -c "${CHANNEL}" -f "${ROOTDIR}/genesisdata/genesis.json" -n "${NETWORK}" -b "${S3_UPLOAD_BUCKET}"
```
Comment (winder) [inline scripts/create_and_deploy_recipe.sh]: Could we have this be configurable in the same way that `create_and_deploy_private_version.sh` is?

**Comment by winder** [comment]
@egieseke below is the test plan from the `S3_UPLOAD_BUCKET` changes, since you've removed `updatekey.json` and extended the cases where `S3_UPLOAD_BUCKET` are used it needs to be updated. Hopefully the scenarios are the same though so we'll just need to make sure they still work correctly.

Notes from https://github.com/algorand/go-algorand-archive/pull/1467
-----
When `-b` is provided, `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` will be used.
When `-b` is not provided:

* Uploads use `S3_UPLOAD_ID`, `S3_UPLOAD_KEY`, and `S3_UPLOAD_BUCKET` will be used for
* Downloads use `updatekey.json` (which now has a `Bucket` field).

**Tests**

`with -b`: verified that AWS environment variables are used

* `updater send -s ~/dir -c my_channel -b algorand-testing-bucket2`
* `./scripts/deploy_private_version.sh -c willschannel -g willsgenesisnetwork -n devnet -b algorand-testing-bucket2`
* `updater ver check -c stable`
* `updater ver get -c stable -o blah.tar.gz`
* `updater gettools -c willschannel -o tools.tar.gz -b algorand-testing-bucket2`

`without -b` and `upload`: Uses `S3_UPLOAD_*` environment variables.

* `updater send -s ~/dir -c my_channel`
* `./scripts/deploy_private_version.sh -c willschannel -g devnet -n willsnetwork`

`without -b` and `download`: Uses `updatekeys.json`

* `updater ver check -c stable`
* `updater ver get -c willschannel -o blah.tar.gz -b algorand-testing-bucket`
* `updater gettools -c stable -o tools.tar.gz`

`update.sh` tests

* Deploy private version of this branch to `testnet-latest` / buckettest (and `algorand-testing-bucket`).
* Install new update script/tools: `./update.sh -i -c buckettest -d ~/algorand/devnet/data/node/`
* Use new tools to install from specified bucket: `./update.sh -i -c buckettest -d ~/algorand/devnet/data/node/ -b algorand-testing-bucket`
* Use new tools to install from bucket in `updatekey.json`: `./update.sh -i -c nightly -d ~/algorand/devnet/data/node/`

File: cmd/nodecfg/download.go (hunk 8 above)
Comment (egieseke) [reply cmd/nodecfg/download.go]: OK, updated to use credentials for downloading config from a private bucket.

File: util/s3/s3Helper.go (hunk 9 above)
Comment (egieseke) [reply util/s3/s3Helper.go]: Removed the conditional logic from the S3Helper.

File: scripts/create_and_deploy_recipe.sh (hunk 11 above)
Comment (egieseke) [reply scripts/create_and_deploy_recipe.sh]: Updated logic to check either flag or environment variable in both scripts.

File: scripts/create_and_deploy_private_version.sh (hunk 10 above)
Comment (egieseke) [reply scripts/create_and_deploy_private_version.sh]: fixed.

**Comment by Karmastic** [comment]
This has been superseded by #124
